enforcer.go

```diff
@@ -537,13 +545,24 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 	var expression *govaluate.EvaluableExpression
 	hasEval := util.HasEval(expString)
 
-	if hasEval {
+	if !hasEval {
+		var cachedExpression, isPresent = e.expressions.Load(expString)
+		if !isPresent {
+			expression, err = govaluate.NewEvaluableExpressionWithFunctions(expString, functions)
+			if err != nil {
+				return false, err
+			}
+			e.expressions.Store(expString, expression)
+		} else {
+			expression = cachedExpression.(*govaluate.EvaluableExpression)
+		}
+	} else {
```

Comment:
I think the logic of this part of the code should be
1. If hasEval then generateEval function
2.if !Present then new expression
else use cache. 
If hasEval, it may still use cache.

Reply:
When using the cache with 'hasEval', a lot of tests fail... For now I'm restricting the cache to matchers without eval

Reply:
That's weird. Did you find out why?

Reply:
At first glance it looks like the cached key is the matcher-string with the eval not filled in, while the EvaluableExpression has the request parameters filled in. This means that a second call, with different request parameters gets a hit on the matcher string, but uses the Expression with the parameters of the previous request.
I'm not sure how to solve this, but I'm inclined to believe that eval matchers are "worth" to be regenerated into new Expressions each time, because of the added eval-complexity.